Scoped package support for NPM

divido opened the pull request:
This adds support for npm packages that are scoped by the convention `@SCOPE/PACKAGE`, such as `@octokit/rest`. Some of the package parsing regexes relied on the presence of at most one "@" to separate the name from the version. This adds support for multiple "@"s, with the last of them separating the package from the version. As an exception, an "@" at the beginning is not a separator, even if it is the last one.

I added some test cases to show what I mean. Because they hit the npm registry as part of operating, you'll need an active internet connection to run the tests. And, it affects your `~/.cache`. And it takes some time to run -- if you're on a slow connection you could run into problems with the jasmine default timeout value. Up to you whether to keep these as jasmine tests or not.

divido committed:
Add explicit toString() to fs.writeFile data argument

Without this, preparePackageCacheBundle is returning the following error:
TypeError [ERR_INVALID_ARG_TYPE]: The "data" argument must be of type string or an instance of Buffer, TypedArray, or DataView. Received type number (1617392542575)

divido committed:
Adding tests for the npm package

These initial tests do require network access the first time, but after they are cached it will work while offline.

divido committed:
Changing regular expression to allow @ symbols in package names

Some npm packages have @ symbols at the beginning, and tasklemon's regexes that parse package+version combos mistake the initial @
for the version separator. These new regular expressions use the last @ symbol to be the separator, rather than the first.

Note that in the case of package pinning (that is, "#require PACKAGE@VERSION", the @ is mandatory -- this makes the regex
simpler. In the case of direct usage, we assume the @ is part of the package name if it is the first character.

divido committed:
Remove filter that discards text after the slash in npm package names

Some packages use the slash, so we need to support it in the package name

divido commented on source/PackageCache.js:
```diff
@@ -157,7 +157,6 @@ module.exports = {
 	// // Tools
 	_normalizePackageList(packageList, packageVersions = {}) {
 		const normalized = packageList
-			.map(packageName => /[^/]+/.exec(packageName)[0])
```
I'm a bit concerned about this change. It looked like this `.map` call was intentionally added, but it only seems to cause problems from my perspective; and I didn't see any tests fail when I removed it. Was the original purpose overcome by events, or am I missing something important?

Cykelero replied:
This actually is useful! It's for requiring files other than the package's entry point:

```js
#require uuid@3.3.0
const uuid = npm['uuid/v4'];
cli.tell('New unique identifier: ' + uuid());
```

It's pretty uncommon and apparently getting even more rare, but it'd be good to keep supporting it. What kind of issues did this cause?

divido replied:
NPM packages can have slashes embedded in them, like the `@octokit/rest` package in my examples. Truncating the package name causes the wrong name to show up in package.json.

I'll add your snippet as a new test case to support, and try to come up with a way to support both. Do you have ideas or preferences on that? Perhaps an escape character of sorts to indicate that the slash is part of the package name and not a separator for requiring other files.

It may take me a couple days to get back to this.

Cykelero replied:
Oh, I didn't catch that!
Clearly, packages name with slashes in them seem more common than packages than require you to access a sub-file, so we should prioritize the former use case when designing how to distinguish between the two.

An escape character sounds like a good idea. Because of how minimalistic Tasklemon's npm module is, it's really hard to add a way to pass parameters/more info when calling it, so getting the file path to fit nicely in the key would be ideal.
(that's a big part of why version pinning is done through a header line! I tried very hard to allow specifying the version inline, at the same place the package is used, but could never get it to look right)

Looking at npm's [package naming rules](https://www.npmjs.com/package/validate-npm-package-name#naming-rules), non-URL-safe characters are disallowed, so we could use these as separators or escape characters.
A colon separator might be a good option to separate package name from file path:
- `const uuid = npm['uuid:v4']`
- `const exampleTree = npm['@example/tree:v4/es5']`

I'll keep thinking about it, maybe we can do better still, but this already looks pretty good to me!

Cykelero commented on spec/npm/npmSpec.js:
```diff
@@ -0,0 +1,52 @@
+const fs = require('fs');
+
+describe('The npm module', function() {
```
Let's keep this simple, in line with the other describe names.

```suggestion
describe('npm', function() {
```

Cykelero commented on spec/npm/npmSpec.js:
```diff
@@ -0,0 +1,52 @@
+
+	it('should recognize pinned versions for scoped packages', async function() {
+		// The #require must be in column 0 of the script, thus the unusual formatting
+		const scriptSource = `#require @octokit/core@3.0.0
```
Just in case, let's test a more specific version, like `3.2.4`, to catch any issues with parsing the minor and patch version numbers.

divido replied:
Implemented by 963b109

Cykelero commented on spec/npm/npmSpec.js:
```diff
@@ -0,0 +1,52 @@
+		const scriptSource = `
+			const octokit = new npm['@octokit/rest'].Octokit();
+
+			let lemon = (await octokit.rest.repos.listForUser({ username: "Cykelero" }))
```
Accessing the network is inevitable to test npm support, but it'd be good to avoid making an extra network call when checking the package itself, as with `listForUser` here. Do you think you could find a scoped package that doesn't require network access?

divido replied:
Probably. 

However, does this test add any value that isn't covered by `should recognize pinned versions for scoped packages`? When writing, I thought of the pinned versioning after the main fix and added a test for it, but looking at it now, it seems like it also checks the scoped package part. Maybe drop down to just the one test?

Cykelero replied:
I think you were right to create the two tests; this one is less specific, and it might not fail when `recognize pinned versions`, helping narrow down potential issues.

And, thinking of it—since all these tests don't reset the package bundle cache, if all goes well they shouldn't access the network at all, when making use of `npm`, once the cache has been primed. Meaning, if you manage to replace the `listForUser` call with something that doesn't hit the network, the tests will add virtually no time to test runs!

Cykelero reviewed (changes requested):
Dang, my first contributor. Thanks for the two PRs! 😄
The code looks great; it's tidy and minimal, and you've respected the style of the existing code nicely.

Great that you wrote these npm tests, too. The install process is pretty complex, so tests were well-needed.

I have a bit of feedback, but your code is mostly good to go!

Cykelero commented:
And about test speed, you're right, it could be annoying eventually; let's keep these as you wrote them for now, and if they prove to be too much of a slow-down, I'll find a way to split them from the other tests, or make them optional.

divido committed:
Restore the ability to require sub-files, but this time with more colons

In some cases, we need to call require() with a specific file in mind, which needs to be encoded as 'package/file'. Building that
into the npm['...'] syntax is tricky, because the PackageCache needs to separate the package names from the files; and packages can
contain slashes in them.

To make this simple to parse, we have introduced a new symbol, ':' to separate the package name from the file to include. Just
before we actually invoke node's require(), we switch it back to a '/'.

Cykelero commented on docs/api-reference-source/npm.xml:
```diff
@@ -37,10 +37,10 @@ cli.tell('Total count of unique friend names: ' + uniqueFriendNames.length);
-				<df-sample name="Use the uuid package">
-const uuid = npm['uuid/v4'];
-cli.tell('New unique identifier: ' + uuid());
+				
+				<df-sample name="Use the Octokit package">
+const octokit = new npm['@octokit/rest'].Octokit();
+let myRepos = await octokit.rest.repos.listForUser({ username: myName }));
```
And you updated the docs too, perfect 😊

Here I'd like to see if we can make this example even simpler. In all the samples, I try to have as little complexity as possible, as little stuff detracting from what's being illustrated—here, for instance, I'm worried someone looking at the code could wonder whether or not the `new` operator is part of the npm module syntax. If you can find a scoped package that doesn't need instantiating, it'll make the sample even clearer.

Similarly, it'd be better if the actual call to the package was simpler (the whole “await octokit.rest.repos…” thing), but that's a lot less important; I don't expect it to cause confusion, it just adds to the sample size.

divido replied:
This was harder than expected. A lot of packages that are going through the trouble of scoping include multiple types that are intended to be instantiated. I found a derivation of the uuid package (staying on the theme, I guess). What do you think of that one?

I've been using [all-the-package-names](https://www.npmjs.com/package/all-the-packages) `| grep '@' | less` and randomly picking packages to try... :)

Cykelero commented on docs/api-reference-source/npm.xml:
```diff
@@ -70,6 +70,20 @@ $ lemon --pin-packages script.js
 			]]>
 			</discussion>
 		</detail>
+		<detail name="Requiring other files" id="specifying-versions">
```
```suggestion
		<detail name="Requiring other files" id="requiring-sub-files">
```

Cykelero commented on docs/api-reference-source/npm.xml:
```diff
@@ -70,6 +70,20 @@ $ lemon --pin-packages script.js
+			<discussion>
+			<![CDATA[
+				<p>
+					If your script needs a use a specific sub-file of a package, rather than the package's default, you can specify it using a colon to separate the package name from the path.
```
Let's add a bit more context still.

```suggestion
					If your script needs to use a specific sub-file of a package, rather than the package's main file, you can specify it by using a colon to separate the package name from the sub-file path.
```

Cykelero commented on spec/npm/npmSpec.js:
```diff
@@ -22,19 +22,15 @@ describe('npm', function() {
-				.data.find(({ name }) => name == 'tasklemon');
-
-			here.file('output.json').make(true).content = JSON.stringify(lemon);
+			const stryker = new npm['@stryker-mutator/core'].Stryker({ concurrency: 4 });
+			here.file('output').make(true).content = stryker.cliOptions.concurrency;
```
👍

Cykelero commented on spec/npm/npmSpec.js:
```diff
@@ -49,4 +45,23 @@ describe('npm', function() {
+			const uuid_v1 = npm['uuid:v1'];
+			const uuid_v4 = npm['uuid:v4'];
+
+			here.file('output.json').make(true).content = JSON.stringify({
+				v1: uuid_v1.toString(),
+				v4: uuid_v4.toString()
+			});
```
😁✨ 

```suggestion
			here.file('output.json').make(true).content = {
				v1: uuid_v1.toString(),
				v4: uuid_v4.toString()
			};
```

Cykelero reviewed (changes requested):
Just a couple of details left!

Cykelero reviewed (commented):
Sorry for the slow reply; life stuff's been going on.

Allthings's `uuid` package fits the bill perfectly; that's a great find! Everything is in reaaal good shape.

So, the very last thing to do for this PR is a history rewrite, to reduce the commit list to something clearer.
Here, that'd mean mostly a) rebasing on top of master, to remove the merge commit, and b) fixupping the smaller review commits, with the earlier commits they update; to erase the earlier, since-improved code.
`Update Documentation Block Identifier` would be merged into `Restore the ability to require sub-files`, for instance, because it doesn't hold a very interesting change. You decide what's worth keeping.

If interactive rebases aren't your cup of tea (I totally understand!), I can take care of it.
And then we'll merge the PR 😄

(also, I just noticed your commits had abundant context in their extended description—so nice to see 😌)

divido commented:
No worries on the delays -- we all have busy lives. Anyway, after a long delay myself, I was able to do some rebasing today. Hopefully this is what you were looking for.

Cykelero commented:
Yes, that's perfect.
Thank you for your first contribution 😄 I couldn't have asked for a better one!

divido commented:
Could I trouble you to roll a new release including this MR? I'd like to share my TaskLemon scripts using these features with others, and I don't want to make them install from a github reference. Thanks in advance

Cykelero commented:
Of course! Here it is, [version 0.3.0](https://github.com/Cykelero/tasklemon/releases/tag/v0.3.0).

A fun thing: the changes in your PR are the very first to need [backwards compatibility code](https://github.com/Cykelero/tasklemon/commit/7a4fdf68dd1549e2232d9af46adc64cfb2597398)! I'd prepared for this a long time ago, so it's fun to finally make use of the system.

Also, it's super interesting to me that you're sharing scripts with others (and using Tasklemon at all, honestly—I haven't gotten much feedback so far). I'd love to chat with you about that, via email or IM or whatever else, if you're willing :)
